lib/client/conntest/ssh.go

```diff
@@ -49,30 +59,239 @@ func NewSSHConnectionTester(cd services.ConnectionsDiagnostic) *SSHConnectionTes
+	clientConf.SkipLocalAuth = true
+	clientConf.SSHProxyAddr = sshProxyAddr // TODO(marco): remove the next line?
+	clientConf.SSHProxyAddr = webProxyAddr
+	clientConf.Stderr = io.Discard
+	clientConf.Stdin = &bytes.Buffer{}
+	clientConf.Stdout = processStdout
+	clientConf.TLS = clientConfTLS
+	clientConf.TLSRoutingEnabled = tlsRoutingEnabled
+	clientConf.UseKeyPrincipals = true
+	clientConf.Username = currentUser.GetName()
+	clientConf.WebProxyAddr = webProxyAddr
+	clientConf.AddKeysToAgent = client.AddKeysToAgentNo
+
+	teleClient, err := client.NewClient(clientConf)
+	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	// TODO(marco): test ssh connection
-	// - create certificate for user with extra field
-	// - create ssh client using that certificate
-	// - if connection fails because of a network error, a trace must be included indicating that the host is not reachable
-	// - other traces will be added by the Node itself (rbac checks, principal)
-	connectionDiagnostic.SetMessage("dry-run")
-	connectionDiagnostic.SetSuccess(true)
+	ctxWithTimeout, cancelFunc := context.WithTimeout(ctx, req.DialTimeout)
+	defer cancelFunc()
 
-	if err := s.cd.UpdateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+	if err := teleClient.SSH(ctxWithTimeout, []string{"whoami"}, false); err != nil {
+		return s.handleErrFromSSH(ctx, connectionDiagnosticID, req.SSHPrincipal, err, processStdout)
+	}
+
+	connDiag, err := s.clt.AppendDiagnosticTrace(ctx, connectionDiagnosticID, types.NewSuccessTraceConnectionDiagnostic(
+		types.ConnectionDiagnosticTrace_NODE_PRINCIPAL,
+		fmt.Sprintf("%q user exists in target node", req.SSHPrincipal),
+	))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	connDiag.SetMessage(types.DiagnosticMessageSuccess)
+	connDiag.SetSuccess(true)
+
+	if err := s.clt.UpdateConnectionDiagnostic(ctx, connDiag); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return connDiag, nil
+}
+
+func (s SSHConnectionTester) proxyAddrs() (webProxyAddr string, sshProxyAddr string, err error) {
+	proxies, err := s.clt.GetProxies()
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+
+	if len(proxies) == 0 {
+		return "", "", trace.NotFound("cluster has no proxies")
+	}
+
+	parsedAddrs, err := client.ParseProxyHost(proxies[0].GetAddr())
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+
+	return parsedAddrs.WebProxyAddr, parsedAddrs.SSHProxyAddr, nil
+}
+
+func (s SSHConnectionTester) handleErrFromSSH(ctx context.Context, connectionDiagnosticID string, sshPrincipal string, sshError error, processStdout *bytes.Buffer) (types.ConnectionDiagnostic, error) {
+	// Either the node doesn't exist or the user doesn't have access to it.
+	if trace.IsNotFound(sshError) {
+		connDiag, err := s.clt.AppendDiagnosticTrace(ctx, connectionDiagnosticID, types.NewFailedTraceConnectionDiagnostic(
+			types.ConnectionDiagnosticTrace_RBAC_NODE,
+			"Node not found. Ensure the Node exists and your role allows you to access it.",
```

Comment:
This can also mean that teleport service is not running on the node.

Reply:
I'm checking for a `trace.IsNotFound` error.
I just tested by shutting down the teleport process and I got this:

![image](https://user-images.githubusercontent.com/689271/187508368-42cf304d-c871-44df-9b5d-1596f22c651c.png)

Is there a particular scenario that this message would be returned when the teleport service is not running on the node?